Process logs in a separate thread

jamesob opened the pull request:
This changeset renders and flushes log messages in a dedicated thread, preventing the originating thread from blocking (or failing) on fwrite, fflush, or the various string manipulations done on log messages. Originating threads push log messages into a buffer that is continually flushed. 

Benchmarks live in the comments below.

Open questions:
- Do we want this behavior by default? I think so, given that it seems safer and more performant than doing log processing synchronously with Important Stuff.
- Do we want to allow an opt-out configuration option that makes log processing synchronous again?
- The ring buffer implementation allows either overwriting existing data or blocking when at full capacity. In the case of logging, I've set it to block and wait for capacity instead of dropping messages. Does this seem right?

Future work:
- Reconcile this change with https://github.com/bitcoin/bitcoin/pull/13168 in case that's merged first (I have a patch)
- Even when logging is disabled, buffer debug logs without flushing and then handle SIGSEGV et al by flushing the log buffer (idea being to give the user crash diagnostics).

skeees committed:
logging: perform logfile writes in a separate thread

Instead of flushing logs to disk synchronously within the
originating thread, push the log message into a ring buffer to
be consumed and written to disk by a separate, logging-specific
thread.

Co-authored-by: James O'Beirne <james@chaincode.com>

jnewbery commented:
Concept ACK.

My thoughts on your open questions:

- Do we want this behavior by default?
[JN] Yes
- Do we want to allow an opt-out configuration option that makes log processing synchronous again
[JN] Yes
- The ring buffer implementation allows either overwriting existing data or blocking when at full capacity. In the case of logging, I've set it to block and wait for capacity instead of dropping messages. Does this seem right?
[JN] I think there are at least three options here:
    1. overwrite existing data
    2. drop new messages
    3. block
I'd go with (2), but push a final message to the ring buffer saying `Logging buffer full - dropping new log messages`, and then when the ring buffer is accepting messages again, push a message saying `Logging buffer accepting new messages`. That makes it very clear to anyone reading the log that messages were dropped between timestamp A and timestamp B.

practicalswift commented:
Concept ACK!

Agree with @jnewbery regarding the open questions.

Benchmarks would be interesting.

MarcoFalke commented:
Wouldn't blocking be the status-quo and thus preferable for now?

donaloconnor commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,145 @@
+ *
+ * Allows optional blocking behavior instead of overwriting when at capacity.
+ */
+template<typename ET, int S, typename ST=int>
```
Do we really need the size type template parameter? Especially if you are passing an int through as a value. `size_t` makes more sense to me since std::array is size_t and signed sizes (int) also make no sense.

jamesob replied:
Yep, agree. Fixed.

donaloconnor commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,145 @@
+        std::lock_guard<std::mutex> lock(m_lock);
+        return m_size;
+    }
+    inline size_type MaxSize() const { return S; }
```
constexpr ?

jamesob replied:
Done.

donaloconnor commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,145 @@
+    size_type m_size;
+
+    /** Index at which the next insertion will happen. */
+    std::atomic<ST> m_insert_at_idx;
```
it doesn't make sense why having these (`m_insert_at_idx`and `m_read_at_idx`) as std::atomic is necessary if they are protected by the mutex

jamesob replied:
Good point, fixed.

donaloconnor commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,145 @@
+        assert(m_size > 0);
+
+        value_type data = std::move(m_buffer[m_read_at_idx]);
+        m_buffer[m_read_at_idx] = value_type();
```
if the object is movable it will be moved and the value will be empty already but if it's not movable and it's copied then I guess there's no big deal letting the data present? Seems unnecessary to construct an empty object in a free slot again.

donaloconnor commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @arg no_overwrite  If true, block and wait to write instead of overwriting existing data.
+     */
+    size_type PushBack(value_type v, bool no_overwrite = true)
```
I think having a move version of PushBack with const v& ref push back would make more sense to avoid unnecessary copying.
`PushBack(value_type&& v)` and `PushBack(const value_type& v)`

jamesob replied:
Good catch. Shouldn't `const value_type&` bind to both rvalues and lvalues, though?

jamesob replied:
Fixed.

donaloconnor commented on src/logging.cpp:
```diff
@@ -271,3 +274,40 @@ void BCLog::Logger::ShrinkDebugFile()
+
+    void FlushAll()
+    {
+        for (LogArgs s : log_buffer.PopAll()) {
```
`LogArgs& s` to avoid copies

jamesob replied:
Good catch, fixed.

donaloconnor commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,145 @@
+    }
+
+private:
+    std::array<value_type, S> m_buffer;
```
I feel we should use dynamic/heap memory for this for the following reasons:

1. This is only allocated once on start up and is of fixed size.
2. Passing a large value to this array can potentially have 2 problems:
   1. In the stack it can overflow (if used later by someone else)
   2. In global/static data it will grow the executable size. 

Anyway just what I feel. Maybe std::array is okay in this case though. I'm just imagining uses of this class further down the line

jamesob replied:
I'm going to leave this as-is for now since I don't think it's a real problem, but if you have thoughts to the contrary I'm curious to hear.

donaloconnor replied:
Sounds good to me thanks! I went a bit overboard with the comments anyway.

jamesob replied:
No, I thought your feedback was very useful - thanks!

donaloconnor reviewed (changes requested):
Concept ACK

Good idea :)

Just a few things I've spotted.

Regarding your questions:

> Do we want this behavior by default? I think so, given that it seems safer and more performant than doing log processing synchronously with Important Stuff.

Yes

> Do we want to allow an opt-out configuration option that makes log processing synchronous again?

I'm a bit against adding config options for these kind of things. It either should work in all cases or not be used. I think for logging we should stick to one version and not have it user configurable.

> The ring buffer implementation allows either overwriting existing data or blocking when at full capacity. In the case of logging, I've set it to block and wait for capacity instead of dropping messages. Does this seem right?

I'm very against dropping logging data. It shouldn't happen. The threads pushing logs should wait if it does block but the buffer should be big enough that this is rare anyway.

jamesob commented:
I've pushed changes that address @donaloconnor's good feedback (thanks, Donal).

Still trying to decide what to do re: blocking vs. dropping messages; I'm inclined towards the approach that @jnewbery suggests since it seems to be the safest.

skeees commented on src/bitcoind.cpp:
```diff
@@ -191,5 +191,10 @@ int main(int argc, char* argv[])
-    return (AppInit(argc, argv) ? EXIT_SUCCESS : EXIT_FAILURE);
+    int status = AppInit(argc, argv);
+
+    if (!status) {
```
i think if you do shutdown here it will only work for coind but not qt but in init.cpp it will for both? not entirely sure though

skeees commented on src/init.cpp:
```diff
@@ -291,6 +291,7 @@ void Shutdown()
     globalVerifyHandle.reset();
     ECC_Stop();
     LogPrintf("%s: done\n", __func__);
+    async_logging::Shutdown();
```
oh - why do in both? if you are just concerned about capturing all logs, see my comment on the Shutdown() method below

skeees commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,143 @@
+        if (no_overwrite && m_size == S) {
+            m_read_condvar.wait(lock, [this]() { return this->m_size < S; });
+        }
+        m_buffer[m_insert_at_idx] = v;
```
same thing as in a prior comment - use a forward with an rvalue reference

skeees commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,143 @@
+            {
+                std::unique_lock<std::mutex> lock(m_lock);
+                if (m_size > 0) {
+                    return std::unique_ptr<value_type>(new value_type(AssumeLockedPopFront()));
```
i think you are doing an unnecessary copy here

skeees reviewed (commented):
left some comments - two general ones though:
- you really wrote a thread safe ringbuffer - if you think this will be useful elsewhere, consider having just a pure implementation of the ringbuffer with a derivation (or wrapper) that adds the synchronization logic
- Some of your api naming/consistency is a bit confusing. You have:
`PushBack()` - might block if no_overwrite is set
`PopFront()` - never blocks - user must know data is available
`PollForOne()` - blocks, might return null

On the block vs drop debate - I think I would vote in the blocking camp since that is essentially current behavior, followed by dropping older messages

jamesob commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,143 @@
+     *
+     * @return false if the poll was interrupted without executing func.
+     */
+    std::unique_ptr<value_type> PollForOne(int poll_interval_ms = 200)
```
Ah yeah, that totally makes sense.

jamesob commented on src/ringbuffer.h:
```diff
@@ -0,0 +1,143 @@
+public:
+    using value_type = ET;
+
+    RingBuffer() : m_size(0), m_insert_at_idx(0), m_read_at_idx(0), m_stop_waiting(false) { }
```
I don't think that works. Let's say we instantiate a buffer with `MaxSize() == 10` and add 5 elements, making `m_insert_at_idx == 5`. I'd want to start reading from index 0, but your formula gives me `5 + 4 % 10` = 9, which isn't right. Maybe I'm missing something here.

jamesob commented on src/logging.cpp:
```diff
@@ -271,3 +274,44 @@ void BCLog::Logger::ShrinkDebugFile()
         fclose(file);
 }
+
+namespace async_logging {
```
Yeah, fair point.

jamesob commented on src/logging.cpp:
```diff
@@ -271,3 +274,44 @@ void BCLog::Logger::ShrinkDebugFile()
+        std::call_once(flush_logs_thread_started, [](){
+            flush_logs_thread.reset(new std::thread(ConsumeLogs));
+        });
+        log_buffer.PushBack(std::move(str));
```
Ah yeah, thanks.

jamesob commented on src/logging.cpp:
```diff
@@ -271,3 +274,44 @@ void BCLog::Logger::ShrinkDebugFile()
+        log_buffer.PushBack(std::move(str));
+    }
+
+    void Shutdown()
```
Yep, good point.

jamesob commented:
Thanks for the thorough review and good feedback here, @skeees. I'll address most of your comments in an upcoming push.

Sadly, the benchmarks aren't looking good on this one. Either my implementation of the ringbuffer is dog slow or this Concept is eminently NACKable. 

Doing an IBD to height 250,000 from one (chain-synced) virtual machine to another (unsynced) over my local network, this PR performed roughly 77% _slower_ than master `196c5a` (9m31s vs. 16m57s). That's a baffling deficit but I've been able to reproduce it consistently.

I'm going to try applying @TheBlueMatt's previous work (https://github.com/bitcoinfibre/bitcoinfibre/commit/6b6a3aef0663775b63bac7d0aa07ec5fc4eb9fc9) to master and see if that performs any better. If it doesn't, I'm going to just abandon async logging and continue work on a periodic benchmarking system.

skeees commented:
I'd do some profiling / tweaking before you abandon this entirely. I'd think this should be at a minimum equivalent - and potentially substantially faster than the current way things are logged.

skeees commented:
I was curious about the unexpectedly slow performance and tried something slightly different -  https://github.com/skeees/bitcoin/commit/1bf2307b9bf406cc199f3671e856ecd7a582b2a8

its a bit messy, but good enough for benchmarking which seems to suggest it performs slightly better than the current logging, especially if you give it short workloads (e.g. do a reindex w a stopatheight of 150k) - though I imagine this effect gets washed away if you did e.g. a full IBD.

jamesob commented:
Nice, @skeees! That approach performs the best so far in terms of async logging implementations, though it's still roughly on par with master in my benchmark:

```
vagrant@bench:/tmp/2018-05-11-2018-05-asynclog-13145/bitcoin$ rm -rf data && mkdir data && time ./src/bitcoind -datadir=`pwd`/data -conf=`pwd`/bitcoin.conf -debug=all -printtoconsole=0 -connect=0 -addnode=10.0.0.10 -listen=1 -stopatheight=250000
terminate called without an active exception
Aborted (core dumped)

real    9m37.566s
user    8m27.668s
sys     3m38.395s
```

(the core dump is just an unclean shutdown of the logging thread)

@skeees suggested some ideas for further improvements on his draft which he or I may fool around with.

jamesob commented:
I think I'm giving up the ghost on this one. I've taken @skeees' latest work, fixed shutdown, and simplified the code a bit, but I still can't get benchmarks to show improvement.

Below is a round of three reindexes to height 100,000:

### master (https://github.com/bitcoin/bitcoin/commit/196c5a947a071f7c9c5b211f0a47787a7d540ecb)

```sh
vagrant@bench:/tmp/2018-05-09-master-6186a/bitcoin$ for _ in `seq 3`; do time ./src/bitcoind -datadir=`pwd`/data -conf=`pwd`/bitcoin.conf -debug=all
-printtoconsole=0 -connect=0 -addnode=10.0.0.10 -listen=1 -stopatheight=100000 -reindex -logasync=0; done

real    1m21.080s
user    0m33.459s
sys     0m37.487s

real    1m4.716s
user    0m34.033s
sys     0m33.567s

real    1m23.858s
user    0m34.043s
sys     0m36.156s
```

### this branch  (https://github.com/bitcoin/bitcoin/pull/13200/commits/34000b1aaa626f1547ad80693e5b636857426f65)

```sh
vagrant@bench:/tmp/2018-05-11-2018-05-asynclog-13145/bitcoin$ for _ in `seq 3`; do time ./src/bitcoind -datadir=`pwd`/data -conf=`pwd`/bitcoin.conf -debug=all -printtoconsole=0 -connect=0 -addnode=10.0.0.10 -listen=1 -stopatheight=100000 -reindex; done

real    1m4.269s
user    0m31.258s
sys     0m33.798s

real    1m30.629s
user    0m31.691s
sys     0m40.961s

real    1m14.528s
user    0m32.676s
sys     0m38.040s
```

### skeees version (https://github.com/skeees/bitcoin/commit/57dc3fcdc0b90fa4f0bff004eed8ca5ae21ad6bd)

```sh
vagrant@bench:/tmp/2018-05-11-2018-05-asynclog-13145/bitcoin$ for _ in `seq 3`; do time ./src/bitcoind -datadir=`pwd`/data -conf=`pwd`/bitcoin.conf -debug=all -printtoconsole=0 -connect=0 -addnode=10.0.0.10 -listen=1 -stopatheight=100000 -reindex; done
terminate called without an active exception
Aborted (core dumped)

real    1m18.180s
user    0m33.764s
sys     0m39.323s
terminate called without an active exception
Aborted (core dumped)

real    3m17.639s
user    0m45.444s
sys     2m21.804s
terminate called without an active exception
Aborted (core dumped)

real    4m59.937s
user    0m48.212s
sys     3m35.134s
```

Can't tell what's going on with those last two runs.

In any case, I don't think the numbers here justify the complexity of this changeset. It's surprising to me that we aren't seeing more improvement, but I've got other stuff to do. I'm going to abandon this PR for now, but if anyone wants to pick it up they're welcome to.

practicalswift commented:
I would like to thank you for doing this experimentation @jamesob. Negative results are important too. Thanks!

jamesob commented:
Reopening because @skeees has some interesting bench results...

skeees commented:
wrote a quick benchmarking script here: https://gist.github.com/skeees/93d3ab3ce8ceabfa1b68df51d3b36b60
It compares the different logging approaches mentioned above - specifically a reindex and and ibd up to a fixed height

Some sample results below (we've now had a couple of independent reproductions across a couple of different machines) - and the latest commit on this pr is consistently faster than existing implementations for microbenchmarks (100k blocks) and equivalent if you change height in the script to a larger number (e.g. if you set `HEIGHT=250000` in the script)

I think, given the below, there's decent evidence that this will improve performance for bursty things, and is worth pursuing. A couple things still need to be cleaned up (shutdown mostly)

(total is wall clock time)
```
benchmarking reindex _pin-1805-asynclog-master
59.65s user 7.97s system 197% cpu 34.296 total
59.55s user 7.88s system 198% cpu 33.963 total
benchmarking reindex _pin-1805-asynclog-bluematt
62.03s user 10.91s system 224% cpu 32.530 total
62.88s user 11.10s system 227% cpu 32.569 total
benchmarking reindex _pin-1805-asynclog-skeees-jamesob
57.02s user 5.61s system 232% cpu 26.952 total
56.61s user 5.49s system 232% cpu 26.750 total

benchmarking ibd _pin-1805-asynclog-master
68.38s user 11.55s system 175% cpu 45.427 total
68.25s user 11.60s system 182% cpu 43.749 total
benchmarking ibd _pin-1805-asynclog-bluematt
72.10s user 15.20s system 207% cpu 42.017 total
72.34s user 15.10s system 207% cpu 42.170 total
benchmarking ibd _pin-1805-asynclog-skeees-jamesob
65.20s user 8.46s system 210% cpu 35.001 total
65.10s user 8.88s system 209% cpu 35.307 total
```

practicalswift commented:
@skeees Is `_pin-1805-asynclog-master` the base case – i.e. before this PR?

jamesob commented:
Benchmarking with bitcoinperf shows a speedup of 4.6% on IBD up to height 200,000 with this change ([details here](https://gist.github.com/jamesob/b630242d28287050628f8685934a5c4d)).

gmaxwell commented:
>  I'd go with (2), but push a final message to the ring buffer saying Logging buffer full - dropping new log messages, and then when the ring buffer is accepting messages again, push a message saying Logging buffer accepting new messages

How about (4)  Store a sequence number,  always overwrite.  When reading out, notice when there were missed messages via gaps in the sequence number and log "X missed log entries".

jamesob commented:
> If you're doing a lot of this sort of testing, its useful to bring a node up to height 450k, stop it, back it up and benchmark reindexing up to 500k.

That sounds much more reasonable. I'll rerun benches on that part of the chain in the next few weeks.

elichai commented:
Just a thought I had,
idk if there's lock contention here but you could remove locking from this entirely and replace with atomic pointers, and then you swap the buffer pointer with a new empty buffer pointer, that way the effect on the rest of the threads should be at worse a single cache miss

jamesob commented:
> Just a thought I had

I don't plan to put any more work into this, but this might be a good project to pick up for a newish contributor.
